src/environmentd/src/http/sql.rs

```diff
@@ -116,31 +227,198 @@ impl SqlResult {
         }
     }
 
-    fn ok(client: &mut SessionClient, res: ExecuteResponse) -> SqlResult {
+    fn ok(client: &mut SessionClient, tag: Option<String>) -> SqlResult {
         SqlResult::Ok {
-            ok: res.tag(),
+            ok: tag.expect("ok only called on tag-generating results"),
             notices: make_notices(client),
         }
     }
 }
 
-#[derive(Debug, Serialize)]
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(tag = "type", content = "payload")]
+pub enum WebSocketResponse {
+    ReadyForQuery(String),
+    Notice(Notice),
+    Rows(Vec<String>),
+    Row(Vec<serde_json::Value>),
+    CommandComplete(String),
+    Error(String),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
 pub struct Notice {
     message: String,
     severity: String,
 }
 
+/// Trait describing how to transmit a response to a client. HTTP clients
+/// accumulate into a Vec and send all at once. WebSocket clients send each
+/// message as they occur.
+#[async_trait]
+trait ResultSender {
+    /// Adds a result to the client. Returns Err if sending to the client
+    /// produced an error and the server should disconnect. Returns Ok(true) if
+    /// the statement produced an error and should error the transaction, but
+    /// remain connected. Returns Ok(false) if the statement succeeded.
+    async fn add_result(&mut self, res: StatementResult) -> Result<bool, anyhow::Error>;
```

Comment:
Also subjective, but an alternative API could be something along the lines of `Result<Result<(), Error>, Error>`. That may be more explicit that there are multiple layers of errors and also allows the caller to user the statement error if they want.

Reply:
Maybe even `Result<Result<(), ()>, Error>` if we don't care about the inner error.

Reply:
Done.